File: Refactor/web/src/lib/speculum/sessionBindingAuth.ts
import { SessionAuthQueryParam, SessionCacheBustQueryParam } from './constants'

/**
 * Live-session binding auth on the client.
 *
 * HTTP edges (virtual assets, Dom uploads, WT/WS dial) carry the binding token
 * in the reserved query parameter {@link SessionAuthQueryParam}. There is no
 * session-binding cookie — a mirrored site's own `token=` stays opaque upstream
 * query. Hub RPCs keep MessagePack `token` in the body (already explicit).
 *
 * Query surgery for asset URLs is done on the raw string on purpose —
 * `URL`/`URLSearchParams` normalize percent-encoding and parameter order, which
 * would change the virtual-asset key the sidecar materialized the body under.
 */

/** True for URLs served by the Dom Projection serve plane. */
export function isVirtualAssetUrl(url: string): boolean {
  return url.startsWith('/w7s/virtual-') || url.includes('/virtual-')
}

/**
 * Absolutize against the API origin and stamp the reserved auth parameter.
 * Idempotent: re-applying replaces the existing auth parameter instead of
 * appending a second one. Non-virtual URLs are returned untouched.
 */
export function appendSessionAuth(url: string, token: string, assetBaseUrl = ''): string {
  if (!url || !token) return url
  if (!isVirtualAssetUrl(url)) return url
  const base = assetBaseUrl.replace(/\/$/, '')
  const absolute = url.startsWith('http')
    ? url
    : `${base}${url.startsWith('/') ? url : `/${url}`}`
  return setReservedParam(absolute, SessionAuthQueryParam, token)
}

/**
 * Force a fresh fetch of an already-painted stylesheet. Uses a reserved name so
 * the server strips it before the key lookup — an ad-hoc buster would land in
 * the key and miss the materialized asset.
 */
export function appendCacheBust(url: string, value: string | number): string {
  if (!url) return url
  return setReservedParam(url, SessionCacheBustQueryParam, String(value))
}

/** Stamp sessionId + reserved binding token onto a URL (data-plane dial). */
export function appendSessionBindingQuery(
  url: URL,
  sessionId: string,
  token: string,
): URL {
  url.searchParams.set('sessionId', sessionId)
  url.searchParams.set(SessionAuthQueryParam, token)
  return url
}

/** Replace-or-append `name=value`, preserving every other query part verbatim. */
function setReservedParam(url: string, name: string, value: string): string {
  const hashAt = url.indexOf('#')
  const fragment = hashAt >= 0 ? url.slice(hashAt) : ''
  const withoutFragment = hashAt >= 0 ? url.slice(0, hashAt) : url

  const queryAt = withoutFragment.indexOf('?')
  const path = queryAt >= 0 ? withoutFragment.slice(0, queryAt) : withoutFragment
  const rawQuery = queryAt >= 0 ? withoutFragment.slice(queryAt + 1) : ''

  const lowered = name.toLowerCase()
  const kept = rawQuery
    .split('&')
    .filter((part) => part.length > 0)
    .filter((part) => {
      const eq = part.indexOf('=')
      const key = eq >= 0 ? part.slice(0, eq) : part
      return key.toLowerCase() !== lowered
    })

  kept.push(`${name}=${encodeURIComponent(value)}`)
  return `${path}?${kept.join('&')}${fragment}`
}

File: Speculum.Api/Presentation/Sessions/SessionBindingAuth.cs
using System.Diagnostics.CodeAnalysis;
using Microsoft.AspNetCore.Http;
using Speculum.Api.Sessions.Services.Contracts;

namespace Speculum.Api.Presentation.Sessions;

/// <summary>
/// Live-session binding auth for HTTP edges (virtual assets, data-plane dial).
/// <para>
/// The binding token travels in a <b>reserved</b> query parameter
/// (<see cref="QueryParameterName"/>). A mirrored site's own <c>token=</c> is
/// opaque upstream query: it is never read as Speculum auth. There is no session
/// binding cookie — operator auth cookies are unrelated.
/// </para>
/// Pure string helpers so the collision rules are unit-testable without HTTP plumbing.
/// </summary>
public static class SessionBindingAuth
{
    /// <summary>
    /// Reserved auth parameter. Must match the web client's
    /// <c>SessionAuthQueryParam</c>.
    /// </summary>
    public const string QueryParameterName = "speculum-session-token";

    /// <summary>
    /// Reserved client-side cache buster for forced stylesheet reloads. Must match
    /// the web client's <c>SessionCacheBustQueryParam</c>.
    /// </summary>
    public const string CacheBustQueryParameterName = "speculum-cache-bust";

    /// <summary>
    /// Internal escape hatch for cross-origin dev and harness callers. Not a product path.
    /// </summary>
    public const string HeaderName = "X-Speculum-Session-Token";

    private static readonly string[] ReservedParameterNames =
    [
        QueryParameterName,
        CacheBustQueryParameterName,
    ];

    /// <summary>
    /// Picks the session token from the reserved query parameter, falling back to the
    /// header. Returns <c>null</c> when neither carries a usable value.
    /// </summary>
    public static string? SelectToken(string? reservedQueryValue, string? headerValue)
    {
        if (!string.IsNullOrWhiteSpace(reservedQueryValue))
        {
            return reservedQueryValue.Trim();
        }

        return string.IsNullOrWhiteSpace(headerValue) ? null : headerValue.Trim();
    }

    /// <summary>
    /// Reads the binding token from the reserved query parameter, then the header.
    /// </summary>
    public static string? ReadToken(HttpRequest request)
        => SelectToken(
            request.Query[QueryParameterName].ToString(),
            request.Headers[HeaderName].ToString());

    /// <summary>
    /// Resolves a live session from the request binding token.
    /// When <paramref name="expectedSessionId"/> is set, the binding must match that id
    /// (carriers that require <c>sessionId</c> in the query). Otherwise token-only lookup
    /// is used (virtual-asset edges).
    /// </summary>
    public static bool TryAuthorize(
        HttpRequest request,
        ISessionBindingRegistry bindings,
        ILiveSessionService liveSessions,
        [NotNullWhen(true)] out ILiveSession? live,
        [NotNullWhen(true)] out string? token,
        Guid? expectedSessionId = null)
    {
        live = null;
        token = ReadToken(request);
        if (token is null)
        {
            return false;
        }

        if (expectedSessionId is Guid sessionId)
        {
            if (!bindings.TryGetLive(sessionId, token, out _)
                || !liveSessions.TryGet(sessionId, out live))
            {
                live = null;
                return false;
            }

            return true;
        }

        if (!bindings.TryGetLiveByToken(token, out var binding)
            || !liveSessions.TryGet(binding.SessionId, out live))
        {
            live = null;
            return false;
        }

        return true;
    }

    /// <summary>
    /// Removes only Speculum-reserved parameters from a raw query string, preserving
    /// every remaining part verbatim (order, encoding, duplicate names) so the result
    /// matches the key the producer materialized the asset under.
    /// </summary>
    public static string StripReservedFromQuery(string query)
    {
        if (string.IsNullOrEmpty(query) || query == "?")
        {
            return "";
        }

        var raw = query.StartsWith('?') ? query[1..] : query;
        var kept = raw
            .Split('&', StringSplitOptions.RemoveEmptyEntries)
            .Where(part => !IsReserved(part))
            .ToArray();

        return kept.Length == 0 ? "" : "?" + string.Join('&', kept);
    }

    /// <summary>
    /// Replace-or-append a reserved query parameter without normalizing encoding/order
    /// of other params (matches client <c>appendSessionAuth</c>).
    /// </summary>
    public static string SetReservedParam(string url, string name, string value)
    {
        ArgumentNullException.ThrowIfNull(url);
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(value);

        var hashAt = url.IndexOf('#');
        var fragment = hashAt >= 0 ? url[hashAt..] : "";
        var withoutFragment = hashAt >= 0 ? url[..hashAt] : url;

        var queryAt = withoutFragment.IndexOf('?');
        var path = queryAt >= 0 ? withoutFragment[..queryAt] : withoutFragment;
        var rawQuery = queryAt >= 0 ? withoutFragment[(queryAt + 1)..] : "";

        var kept = rawQuery
            .Split('&', StringSplitOptions.RemoveEmptyEntries)
            .Where(part =>
            {
                var eq = part.IndexOf('=');
                var key = eq >= 0 ? part[..eq] : part;
                return !key.Equals(name, StringComparison.OrdinalIgnoreCase);
            })
            .ToList();

        kept.Add($"{name}={Uri.EscapeDataString(value)}");
        return $"{path}?{string.Join('&', kept)}{fragment}";
    }

    private static bool IsReserved(string queryPart)
    {
        var eq = queryPart.IndexOf('=');
        var name = eq >= 0 ? queryPart[..eq] : queryPart;
        foreach (var reserved in ReservedParameterNames)
        {
            if (name.Equals(reserved, StringComparison.OrdinalIgnoreCase))
            {
                return true;
            }
        }

        return false;
    }
}
